### PR TITLE
Introduces strategies for making bert sentence embeddings

### PR DESCRIPTION
**Pull request by JSybrandt**
Motivation:

  Reading
  https://heartbeat.fritz.ai/research-guide-for-transformers-3ff751493222
  notice how there are multiple different ways to form sentence
  embeddings.  Furthermore, we see challengingly bad knn performance.  I
  would like to understand if we can formulate sentence embeddings in a
  better manner.

  We can now compute sentence embeddings that are averaegd word
  embeddings (as opposed to the pooled layer at the end of BERT).

  Warning: This commit introduces a dependency with pytorch 1.3
  (pytorch 1.3 allows for float * boolean pairwise multiplication)

  Warning: This commit introduces a dependency with transformers
  (module pytorch-transformers has been renamed following a big update).

Features:

  added parser.bert_sentence_method that can be set to "last_layer" or
  "mean_hidden" or "mean_hidden_without_specials" (default). Last_layer
  is what we did before, "mean_hidden" includes all but padding.
  "mean_hidden_without_specials" is everything except padding, as well
  as other specials ([CLS] and [SEP] mostly).

  Embeddings are stored as 16-bit floats.

### CODE REVIEW

File: pymoliere/config/config.proto [hunk 1]
```diff
@@ -82,7 +87,7 @@ message LdaConfig {
 }
 
 message MlConfig {
-  optional int32 batch_size = 1 [default=32];
+  optional int32 batch_size = 1 [default=16];
```
Comment (JSybrandt) [inline pymoliere/config/config.proto]: Encountered memory issues related to storing the mask values.

File: pymoliere/construct/embedding_util.py [hunk 2]
```diff
@@ -1,7 +1,7 @@
 from typing import Tuple, Iterable, ClassVar
 import torch
 import torch
-from pytorch_transformers import (
+from transformers import (
```
Comment (JSybrandt) [inline pymoliere/construct/embedding_util.py]: pytorch_transformers has upgraded to just "transformers"

File: setup.py [hunk 3]
```diff
@@ -94,12 +94,12 @@ def run(self):
       "paramiko",
       "protobuf",
       "pymysql",
-      "pytorch-transformers",
+      "transformers",
```
Comment (JSybrandt) [inline setup.py]: Pytorch-transformers has been renamed to just "transformers"

File: pymoliere/construct/embedding_util.py [hunk 4]
```diff
@@ -86,34 +86,118 @@ def apply_sentence_classifier_to_part(
...
+    mask = mask.unsqueeze(-1).expand_as(embedding)
+
+    # Pairwise multiply the embedding by the mask. The 2'nd to last layer has
+    # one embedding per token in the input. Note that X*false = 0 and X*true = 1
+    embedding *= mask
+
+    # The average is the 1'st axis sum (over the words in the sequence)
+    # pairwise divided by the total number of words (sum of the mask values).
+    embedding = embedding.sum(axis=1)
+    embedding /= mask.sum(axis=1)
```
Comment (JSybrandt) [inline pymoliere/construct/embedding_util.py]: This line makes us depend on pytorch 1.3

**Comment by JSybrandt** [comment]
Determined that this is an overkill and potentially buggy solution.
I have evidence that we should just always pick the mean of the last hidden layer.
